File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int findMaxRob(vector<int>& nums, vector<int>& dp, int ind, int size){
        if(ind >= size){
            return 0;
        }

        if(dp[ind] != -1)return dp[ind];

        int rob = nums[ind] + findMaxRob(nums, dp, ind + 2, size);
        int notRob = findMaxRob(nums, dp, ind + 1, size);

        return dp[ind] = max(rob, notRob);
    }

    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int>dp(n+1, -1);

        return findMaxRob(nums, dp, 0, n);
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int n;


    // private int solve(int[] nums, int ind, int[] dp){
    //     if(ind >= n) return 0;

    //     if(dp[ind] != -1) return dp[ind];

    //     int take = nums[ind] + solve(nums, ind + 2, dp);
    //     int notTake = solve(nums, ind + 1, dp);

    //     return dp[ind] = Math.max(take, notTake);
    // }


    public int rob(int[] nums) {
        n = nums.length;
        int[] house = new int[n + 1];
        
        house[0] = 0;
        house[1] = nums[0];

        for(int i = 2; i <= n; i++){
            int take = nums[i - 1] + house[i - 2];
            int notTake = house[i - 1];

            house[i] = Math.max(take, notTake);
        } 

        return house[n];
    }
}